File: projects/dejavu.py/src/dejavu_language/expr.py
"""Pratt-style expression parser over a CodeToken stream."""

from __future__ import annotations

from typing import Any

from dejavu_language.error import ParseError
from dejavu_language.lexer import lex_code
from dejavu_language.token import CodeToken, CodeTokenKind


class ExprParser:
    def __init__(self, source: str, file: str, base: int, input_str: str) -> None:
        self.source = source
        self.file = file
        self.base = base
        self.tokens = lex_code(input_str, source=source, file=file, base=base)
        for t in self.tokens:
            if t.kind == "CodeEnd":
                raise ParseError(
                    "unexpected `%>` inside expression",
                    file=file,
                    start=base + t.start,
                    length=t.end - t.start,
                )
        self.pos = 0

    def parse(self) -> dict[str, Any]:
        expr = self._parse_pipe()
        if self.pos != len(self.tokens):
            span = self._peek_span()
            raise ParseError(
                "trailing input in expression",
                file=self.file,
                start=span[0],
                length=span[1],
                label="unexpected",
            )
        return expr

    def _peek(self) -> CodeToken | None:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def _peek_kind(self) -> CodeTokenKind | None:
        t = self._peek()
        return t.kind if t else None

    def _peek_span(self) -> tuple[int, int]:
        t = self._peek()
        if t:
            return self.base + t.start, max(1, t.end - t.start)
        last = self.tokens[-1] if self.tokens else None
        end = self.base + (last.end if last else 0)
        return end, 1

    def _bump(self) -> CodeToken | None:
        t = self._peek()
        if t is not None:
            self.pos += 1
        return t

    def _expect_ident(self) -> str:
        t = self._bump()
        if t is not None and t.kind == "Ident":
            return t.text
        span = self._peek_span()
        raise ParseError(
            "expected identifier",
            file=self.file,
            start=span[0],
            length=span[1],
            label="expected ident",
        )

    def _parse_pipe(self) -> dict[str, Any]:
        left = self._parse_or()
        while self._peek_kind() == "PipeOp":
            self._bump()
            filt = self._expect_ident()
            args: list[dict[str, Any]] = []
            if self._peek_kind() == "LParen":
                self._bump()
                if self._peek_kind() != "RParen":
                    while True:
                        args.append(self._parse_pipe())
                        if self._peek_kind() == "Comma":
                            self._bump()
                            continue
                        break
                closed = self._bump()
                if closed is None or closed.kind != "RParen":
                    span = self._peek_span()
                    raise ParseError(
                        "expected `)` after filter arguments",
                        file=self.file,
                        start=span[0],
                        length=span[1],
                    )
            left = {
                "type": "Expr.Pipe",
                "expression": left,
                "filter": filt,
                "arguments": args,
            }
        return left

    def _parse_or(self) -> dict[str, Any]:
        left = self._parse_and()
        while self._peek_kind() == "OrOr":
            self._bump()
            left = {
                "type": "Expr.Binary",
                "operator": "||",
                "left": left,
                "right": self._parse_and(),
            }
        return left

    def _parse_and(self) -> dict[str, Any]:
        left = self._parse_cmp()
        while self._peek_kind() == "AndAnd":
            self._bump()
            left = {
                "type": "Expr.Binary",
                "operator": "&&",
                "left": left,
                "right": self._parse_cmp(),
            }
        return left

    def _parse_cmp(self) -> dict[str, Any]:
        left = self._parse_add()
        kind = self._peek_kind()
        op_map: dict[CodeTokenKind, str] = {
            "EqEq": "==",
            "NotEq": "!=",
            "LessEq": "<=",
            "GreaterEq": ">=",
            "Less": "<",
            "Greater": ">",
            "In": "in",
        }
        op = op_map.get(kind) if kind else None
        if op is not None:
            self._bump()
            return {
                "type": "Expr.Binary",
                "operator": op,
                "left": left,
                "right": self._parse_add(),
            }
        return left

    def _parse_add(self) -> dict[str, Any]:
        left = self._parse_mul()
        while True:
            kind = self._peek_kind()
            if kind in ("Plus", "Minus"):
                op = "+" if kind == "Plus" else "-"
                self._bump()
                left = {
                    "type": "Expr.Binary",
                    "operator": op,
                    "left": left,
                    "right": self._parse_mul(),
                }
            else:
                break
        return left

    def _parse_mul(self) -> dict[str, Any]:
        left = self._parse_unary()
        while True:
            kind = self._peek_kind()
            if kind in ("Star", "Slash", "Percent"):
                op = "*" if kind == "Star" else "/" if kind == "Slash" else "%"
                self._bump()
                left = {
                    "type": "Expr.Binary",
                    "operator": op,
                    "left": left,
                    "right": self._parse_unary(),
                }
            else:
                break
        return left

    def _parse_unary(self) -> dict[str, Any]:
        kind = self._peek_kind()
        if kind in ("Bang", "Minus", "Plus"):
            op = "!" if kind == "Bang" else "-" if kind == "Minus" else "+"
            self._bump()
            return {
                "type": "Expr.Unary",
                "operator": op,
                "argument": self._parse_unary(),
            }
        return self._parse_postfix()

    def _parse_postfix(self) -> dict[str, Any]:
        left = self._parse_primary()
        while True:
            kind = self._peek_kind()
            if kind == "Dot":
                self._bump()
                left = {
                    "type": "Expr.Member",
                    "object": left,
                    "property": self._expect_ident(),
                }
            elif kind == "LBracket":
                self._bump()
                index = self._parse_pipe()
                closed = self._bump()
                if closed is None or closed.kind != "RBracket":
                    span = self._peek_span()
                    raise ParseError(
                        "expected `]`",
                        file=self.file,
                        start=span[0],
                        length=span[1],
                    )
                left = {"type": "Expr.Index", "object": left, "index": index}
            elif kind == "LParen":
                self._bump()
                args: list[dict[str, Any]] = []
                if self._peek_kind() != "RParen":
                    while True:
                        args.append(self._parse_pipe())
                        if self._peek_kind() == "Comma":
                            self._bump()
                            continue
                        break
                closed = self._bump()
                if closed is None or closed.kind != "RParen":
                    span = self._peek_span()
                    raise ParseError(
                        "expected `)`",
                        file=self.file,
                        start=span[0],
                        length=span[1],
                    )
                left = {"type": "Expr.Call", "callee": left, "arguments": args}
            else:
                break
        return left

    def _parse_primary(self) -> dict[str, Any]:
        t = self._bump()
        if t is None:
            span = self._peek_span()
            raise ParseError(
                "unexpected end of expression",
                file=self.file,
                start=span[0],
                length=span[1],
            )
        if t.kind == "String":
            return {"type": "Expr.Literal", "value": t.text}
        if t.kind == "Bool":
            return {"type": "Expr.Literal", "value": bool(t.value)}
        if t.kind == "Null":
            return {"type": "Expr.Literal", "value": None}
        if t.kind == "Number":
            try:
                if "." in t.text:
                    num: int | float = float(t.text)
                else:
                    num = int(t.text)
            except ValueError as e:
                raise ParseError(
                    f"invalid number `{t.text}`",
                    file=self.file,
                    start=self.base + t.start,
                    length=t.end - t.start,
                ) from e
            return {"type": "Expr.Literal", "value": num}
        if t.kind == "Ident":
            return {"type": "Expr.Identifier", "name": t.text}
        if t.kind == "LParen":
            e = self._parse_pipe()
            closed = self._bump()
            if closed is None or closed.kind != "RParen":
                span = self._peek_span()
                raise ParseError(
                    "expected `)`",
                    file=self.file,
                    start=span[0],
                    length=span[1],
                )
            return e
        raise ParseError(
            "unexpected token in expression",
            file=self.file,
            start=self.base + t.start,
            length=max(1, t.end - t.start),
        )


def parse_expr(
    input_str: str,
    *,
    source: str | None = None,
    file: str = "template.dejavu",
    base: int = 0,
) -> dict[str, Any]:
    src = source if source is not None else input_str
    return ExprParser(src, file, base, input_str).parse()

File: projects/dejavu.ts/dejavu-language/src/expr.ts
import type {IrNode} from "@dejavu/types";
import {ParseError} from "./error";
import {lexCode} from "./lexer";
import type {CodeToken, CodeTokenKind} from "./token";

/** Pratt-style expression parser over a [`CodeToken`] stream. */
export class ExprParser {
    private readonly source: string;
    private readonly file: string;
    private readonly base: number;
    private readonly tokens: CodeToken[];
    private pos = 0;

    constructor(source: string, file: string, base: number, input: string) {
        this.source = source;
        this.file = file;
        this.base = base;
        this.tokens = lexCode(input, {source, file, base});
        for (const t of this.tokens) {
            if (t.kind === "CodeEnd") {
                throw new ParseError("unexpected `%>` inside expression", {
                    file,
                    start: base + t.start,
                    length: t.end - t.start,
                });
            }
        }
    }

    parse(): IrNode {
        const expr = this.parsePipe();
        if (this.pos !== this.tokens.length) {
            const span = this.peekSpan();
            throw new ParseError("trailing input in expression", {
                file: this.file,
                start: span.start,
                length: span.length,
                label: "unexpected",
            });
        }
        return expr;
    }

    private peek(): CodeToken | undefined {
        return this.tokens[this.pos];
    }

    private peekKind(): CodeTokenKind | undefined {
        return this.tokens[this.pos]?.kind;
    }

    private peekSpan(): { start: number; length: number } {
        const t = this.tokens[this.pos];
        if (t) return {start: this.base + t.start, length: Math.max(1, t.end - t.start)};
        const last = this.tokens.at(-1);
        const end = this.base + (last?.end ?? 0);
        return {start: end, length: 1};
    }

    private bump(): CodeToken | undefined {
        return this.tokens[this.pos++];
    }

    private expectIdent(): string {
        const t = this.bump();
        if (t?.kind === "Ident") return t.text;
        const span = this.peekSpan();
        throw new ParseError("expected identifier", {
            file: this.file,
            start: span.start,
            length: span.length,
            label: "expected ident",
        });
    }

    private parsePipe(): IrNode {
        let left = this.parseOr();
        while (this.peekKind() === "PipeOp") {
            this.bump();
            const filter = this.expectIdent();
            const args: IrNode[] = [];
            if (this.peekKind() === "LParen") {
                this.bump();
                if (this.peekKind() !== "RParen") {
                    for (; ;) {
                        args.push(this.parsePipe());
                        if (this.peekKind() === "Comma") {
                            this.bump();
                            continue;
                        }
                        break;
                    }
                }
                if (this.bump()?.kind !== "RParen") {
                    const span = this.peekSpan();
                    throw new ParseError("expected `)` after filter arguments", {
                        file: this.file,
                        start: span.start,
                        length: span.length,
                    });
                }
            } else if (this.peekKind() === "Colon") {
                // Liquid / Doki: `value | date: '%Y-%m-%d'` / `value | default: x`
                this.bump();
                args.push(this.parseOr());
                while (this.peekKind() === "Comma") {
                    this.bump();
                    args.push(this.parseOr());
                }
            }
            left = {type: "Expr.Pipe", expression: left, filter, arguments: args};
        }
        return left;
    }

    private parseOr(): IrNode {
        let left = this.parseAnd();
        while (this.peekKind() === "OrOr") {
            this.bump();
            left = {type: "Expr.Binary", operator: "||", left, right: this.parseAnd()};
        }
        return left;
    }

    private parseAnd(): IrNode {
        let left = this.parseCmp();
        while (this.peekKind() === "AndAnd") {
            this.bump();
            left = {type: "Expr.Binary", operator: "&&", left, right: this.parseCmp()};
        }
        return left;
    }

    private parseCmp(): IrNode {
        const left = this.parseAdd();
        const kind = this.peekKind();
        const op =
            kind === "EqEq"
                ? "=="
                : kind === "NotEq"
                    ? "!="
                    : kind === "LessEq"
                        ? "<="
                        : kind === "GreaterEq"
                            ? ">="
                            : kind === "Less"
                                ? "<"
                                : kind === "Greater"
                                    ? ">"
                                    : kind === "In"
                                        ? "in"
                                        : null;
        if (op) {
            this.bump();
            return {type: "Expr.Binary", operator: op, left, right: this.parseAdd()};
        }
        return left;
    }

    private parseAdd(): IrNode {
        let left = this.parseMul();
        for (; ;) {
            const kind = this.peekKind();
            if (kind === "Plus" || kind === "Minus") {
                const op = kind === "Plus" ? "+" : "-";
                this.bump();
                left = {type: "Expr.Binary", operator: op, left, right: this.parseMul()};
            } else break;
        }
        return left;
    }

    private parseMul(): IrNode {
        let left = this.parseUnary();
        for (; ;) {
            const kind = this.peekKind();
            if (kind === "Star" || kind === "Slash" || kind === "Percent") {
                const op = kind === "Star" ? "*" : kind === "Slash" ? "/" : "%";
                this.bump();
                left = {type: "Expr.Binary", operator: op, left, right: this.parseUnary()};
            } else break;
        }
        return left;
    }

    private parseUnary(): IrNode {
        const kind = this.peekKind();
        if (kind === "Bang" || kind === "Minus" || kind === "Plus") {
            const op = kind === "Bang" ? "!" : kind === "Minus" ? "-" : "+";
            this.bump();
            return {type: "Expr.Unary", operator: op, argument: this.parseUnary()};
        }
        return this.parsePostfix();
    }

    private parsePostfix(): IrNode {
        let left = this.parsePrimary();
        for (; ;) {
            const kind = this.peekKind();
            if (kind === "Dot") {
                this.bump();
                left = {type: "Expr.Member", object: left, property: this.expectIdent()};
            } else if (kind === "LBracket") {
                this.bump();
                const index = this.parsePipe();
                if (this.bump()?.kind !== "RBracket") {
                    const span = this.peekSpan();
                    throw new ParseError("expected `]`", {
                        file: this.file,
                        start: span.start,
                        length: span.length,
                    });
                }
                left = {type: "Expr.Index", object: left, index};
            } else if (kind === "LParen") {
                this.bump();
                const args: IrNode[] = [];
                if (this.peekKind() !== "RParen") {
                    for (; ;) {
                        args.push(this.parsePipe());
                        if (this.peekKind() === "Comma") {
                            this.bump();
                            continue;
                        }
                        break;
                    }
                }
                if (this.bump()?.kind !== "RParen") {
                    const span = this.peekSpan();
                    throw new ParseError("expected `)`", {
                        file: this.file,
                        start: span.start,
                        length: span.length,
                    });
                }
                left = {type: "Expr.Call", callee: left, arguments: args};
            } else break;
        }
        return left;
    }

    private parsePrimary(): IrNode {
        const t = this.bump();
        if (!t) {
            const span = this.peekSpan();
            throw new ParseError("unexpected end of expression", {
                file: this.file,
                start: span.start,
                length: span.length,
            });
        }
        switch (t.kind) {
            case "String":
                return {type: "Expr.Literal", value: t.text};
            case "Bool":
                return {type: "Expr.Literal", value: Boolean(t.value)};
            case "Null":
                return {type: "Expr.Literal", value: null};
            case "Number": {
                const num = Number(t.text);
                if (Number.isNaN(num)) {
                    throw new ParseError(`invalid number \`${t.text}\``, {
                        file: this.file,
                        start: this.base + t.start,
                        length: t.end - t.start,
                    });
                }
                return {type: "Expr.Literal", value: num};
            }
            case "Ident":
                return {type: "Expr.Identifier", name: t.text};
            case "LParen": {
                const e = this.parsePipe();
                if (this.bump()?.kind !== "RParen") {
                    const span = this.peekSpan();
                    throw new ParseError("expected `)`", {
                        file: this.file,
                        start: span.start,
                        length: span.length,
                    });
                }
                return e;
            }
            default: {
                throw new ParseError("unexpected token in expression", {
                    file: this.file,
                    start: this.base + t.start,
                    length: Math.max(1, t.end - t.start),
                });
            }
        }
    }
}

export function parseExpr(
    input: string,
    opts?: { source?: string; file?: string; base?: number },
): IrNode {
    const source = opts?.source ?? input;
    const file = opts?.file ?? "template.dejavu";
    const base = opts?.base ?? 0;
    return new ExprParser(source, file, base, input).parse();
}
